File: frontend/src/lib/api/trainee.ts
/**
 * Praktykanci — codzienne listy telefonów (0374).
 *
 * Typy są lustrem kontraktu `docs/trainee-call-lists-contract.md`
 * (`backend/app/api/trainee.py`). Klucze react-query są eksportowane, żeby
 * harnessy `/preview/trainee*` mogły zasiać cache tymi samymi funkcjami.
 * Kwoty stawek są zawsze PLN netto B2B.
 */

import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

import { api } from "@/lib/api";
import { recountToday } from "@/lib/trainee-call";

// ── Praktykant: lista na dziś ────────────────────────────────────────────────

export type TraineeListStatus =
  | "ready"
  | "not_workday"
  | "no_program"
  | "program_finished"
  /** Admin w „podglądzie jako”: lista na dziś jeszcze nie powstała, a podgląd
   *  jej nie zakłada (tylko odczyt). */
  | "preview_not_generated";

/** Wynik zamkniętej pozycji. `call` = rozmowa zapisana formularzem. */
export type TraineeOutcome = "call" | "noanswer" | "later" | "wrong" | "declined";

export type TraineeMissingCode =
  | "rate_missing"
  | "rate_stale"
  | "b2b"
  | "work_time"
  | "work_mode"
  | "below_min_consent"
  | "office_consent"
  | "availability";

export type B2bWillingness = "b2b" | "would_switch" | "employment_only";
export type RemoteMode = "remote" | "hybrid" | "onsite";
export type WorkTimePreference =
  | "full_time_only"
  | "also_part_time"
  | "part_time_only";
export type CallAvailability = "now" | "within_1m" | "within_3m" | "later";
export type OpenToOffers = "yes" | "maybe" | "no";
export type RateUnit = "hour" | "day" | "month";

export interface TraineeItemFacts {
  min_rate_hourly: number | null;
  rate_updated_at: string | null;
  b2b_willingness: B2bWillingness | null;
  accepts_below_min_rate: boolean | null;
  remote_modes: RemoteMode[];
  max_onsite_days: number | null;
  accepts_more_office_days: boolean | null;
  office_cities: string[];
  work_time_preference: WorkTimePreference | null;
  availability_status: string | null;
  availability_date: string | null;
}

export interface TraineeItemReasons {
  /** Do ilu rekrutacji z okna reguł pasuje. */
  fits: number;
  /** …z czego ile jest otwartych teraz. */
  open_fits: number;
  stack: string[];
  missing: TraineeMissingCode[];
}

export interface TraineeItem {
  id: number;
  candidate_id: number;
  position: number;
  name: string;
  role: string | null;
  company: string | null;
  city: string | null;
  phone: string | null;
  in_base_since: number | null;
  last_contact_at: string | null;
  attempts: number;
  /** Pierwsza próba bez odpowiedzi: wraca na listę po tej chwili. */
  retry_after: string | null;
  outcome: TraineeOutcome | null;
  closed_at: string | null;
  later_date: string | null;
  reasons: TraineeItemReasons;
  facts: TraineeItemFacts;
}

export interface TraineeProgramDay {
  day: number;
  total_days: number;
  status: string;
}

export interface TraineeCounts {
  total: number;
  closed: number;
  open: number;
  call: number;
  noanswer: number;
  later: number;
  wrong: number;
  declined: number;
}

export interface TraineeToday {
  list_date: string;
  status: TraineeListStatus;
  program: TraineeProgramDay | null;
  counts: TraineeCounts;
  day_completed: boolean;
  answered_pct: number | null;
  team_answered_pct: number | null;
  items: TraineeItem[];
}

export interface TraineeCallBody {
  b2b_willingness: B2bWillingness;
  min_rate: { value: number; unit: RateUnit } | null;
  accepts_below_min_rate: boolean | null;
  remote_modes: RemoteMode[];
  max_onsite_days: number | null;
  accepts_more_office_days: boolean | null;
  office_cities: string[];
  work_time_preference: WorkTimePreference | null;
  availability: CallAvailability | null;
  open_to_offers: OpenToOffers | null;
  wants: string | null;
}

export type TraineeNoCallOutcome = Exclude<TraineeOutcome, "call">;

export interface TraineeOutcomeBody {
  outcome: TraineeNoCallOutcome;
  /** Tylko przy `later` — polski dzień roboczy po dziś. */
  later_date?: string;
}

export interface TraineeItemResponse {
  item: TraineeItem;
  day_completed: boolean;
}

export interface TraineeOpenJob {
  job_id: number;
  title: string;
  recruiter_name: string | null;
  matched_skills: string[];
}

export interface TraineeHandoverBody {
  job_id: number;
  note: string;
}

// ── Head of Recruitment / admin ──────────────────────────────────────────────

export interface TraineeProgram {
  start_date: string;
  workdays: number;
  extended_days: number;
  daily_list_size: number;
  status: string;
  day: number;
  total_days: number;
  end_date: string | null;
  decision_due: boolean;
}

export interface TraineeOverviewRow {
  user_id: number;
  name: string;
  is_active: boolean;
  program: TraineeProgram | null;
  days_with_list: number;
  days_completed: number;
  calls_per_day: number | null;
  answered_pct: number | null;
  complete_profiles_pct: number | null;
  handed_over: number;
  handed_in_process: number;
  quality: { checked: number; issues: number };
  flag_low_answer: boolean;
}

export interface TraineePoolCategory {
  name: string;
  count: number;
}

export interface TraineePool {
  size: number;
  open_fit: number;
  by_category: TraineePoolCategory[];
  computed_at?: string | null;
}

export interface TraineeOverview {
  trainees: TraineeOverviewRow[];
  team_answered_pct: number | null;
  /** `null`, dopóki pula nie była jeszcze policzona (pierwszy bieg po wdrożeniu). */
  pool: TraineePool | null;
  month: { verified_rates: number; handed_over: number; in_process: number };
}

export interface TraineeProgramUpdate {
  start_date?: string;
  workdays?: number;
  daily_list_size?: number;
}

export type TraineeDecisionBody =
  | { action: "promote"; role: "sourcer" | "recruiter"; add_to_my_people: boolean }
  | { action: "extend"; extend_days: number }
  | { action: "end" };

export type QualityVerdict = "ok" | "issue";

export interface TraineeQualitySampleItem {
  item_id: number;
  candidate_id: number;
  name: string;
  phone: string | null;
  called_at: string | null;
  facts: Partial<TraineeItemFacts>;
  verdict: QualityVerdict | null;
  note: string | null;
}

export interface TraineeRules {
  min_fits: number;
  window_months: number;
  rate_stale_months: number;
  verified_recently_days: number;
  process_active_days: number;
  my_people_contact_days: number;
  trainee_recall_days: number;
  missing_rate: boolean;
  missing_availability: boolean;
  missing_work_mode: boolean;
  missing_consents: boolean;
  missing_b2b: boolean;
  missing_work_time: boolean;
}

export type TraineeRulesPreview = TraineePool;

// ── Klucze ───────────────────────────────────────────────────────────────────

export const traineeKeys = {
  all: ["trainee"] as const,
  today: () => ["trainee", "today"] as const,
  openJobs: (itemId: number) => ["trainee", "open-jobs", itemId] as const,
  overview: () => ["trainee", "overview"] as const,
  qualitySample: (userId: number) => ["trainee", "quality-sample", userId] as const,
  rules: () => ["trainee", "rules"] as const,
  rulesPreview: (rules: TraineeRules | null) =>
    ["trainee", "rules-preview", rules] as const,
};

// ── Wywołania ────────────────────────────────────────────────────────────────

export const traineeApi = {
  async today() {
    const { data } = await api.get<TraineeToday>("/api/trainee/today");
    return data;
  },
  async saveCall(itemId: number, body: TraineeCallBody) {
    const { data } = await api.post<TraineeItemResponse>(
      `/api/trainee/items/${itemId}/call`,
      body,
    );
    return data;
  },
  async saveOutcome(itemId: number, body: TraineeOutcomeBody) {
    const { data } = await api.post<TraineeItemResponse>(
      `/api/trainee/items/${itemId}/outcome`,
      body,
    );
    return data;
  },
  async openJobs(itemId: number) {
    const { data } = await api.get<TraineeOpenJob[]>(
      `/api/trainee/items/${itemId}/open-jobs`,
    );
    return data;
  },
  async handover(itemId: number, body: TraineeHandoverBody) {
    const { data } = await api.post<{ ok: boolean }>(
      `/api/trainee/items/${itemId}/handover`,
      body,
    );
    return data;
  },
  async overview() {
    const { data } = await api.get<TraineeOverview>("/api/trainee/overview");
    return data;
  },
  async updateProgram(userId: number, body: TraineeProgramUpdate) {
    const { data } = await api.put(`/api/trainee/programs/${userId}`, body);
    return data;
  },
  async decision(userId: number, body: TraineeDecisionBody) {
    const { data } = await api.post(`/api/trainee/programs/${userId}/decision`, body);
    return data;
  },
  async qualitySample(userId: number) {
    const { data } = await api.get<TraineeQualitySampleItem[]>(
      "/api/trainee/quality-sample",
      { params: { user_id: userId } },
    );
    return data;
  },
  async qualityVerdict(itemId: number, body: { verdict: QualityVerdict; note: string }) {
    const { data } = await api.post(`/api/trainee/quality-sample/${itemId}`, body);
    return data;
  },
  async rules() {
    const { data } = await api.get<TraineeRules>("/api/trainee/rules");
    return data;
  },
  async saveRules(body: TraineeRules) {
    const { data } = await api.put<TraineeRules>("/api/trainee/rules", body);
    return data;
  },
  /**
   * Podgląd puli. Bieżący (niezapisany) stan formularza idzie w parametrach,
   * żeby podgląd był „na żywo”; bez nich backend liczy zapisane reguły.
   */
  async rulesPreview(rules: TraineeRules | null) {
    const { data } = await api.get<TraineeRulesPreview>("/api/trainee/rules/preview", {
      params: rules ?? undefined,
    });
    return data;
  },
};

// ── Hooki ────────────────────────────────────────────────────────────────────

export function useTraineeToday(enabled = true) {
  return useQuery({
    queryKey: traineeKeys.today(),
    queryFn: () => traineeApi.today(),
    enabled,
    // Lista zmienia się tylko przez zapisy praktykanta; nocą powstaje nowa.
    staleTime: 60_000,
  });
}

/**
 * Po zapisie pozycji: podmiana wiersza w cache (lista od razu przeskakuje na
 * następną osobę), przeliczenie liczników lokalnie i odświeżenie z serwera.
 */
function useApplyItemResponse() {
  const queryClient = useQueryClient();
  return (response: TraineeItemResponse) => {
    queryClient.setQueryData<TraineeToday>(traineeKeys.today(), (prev) => {
      if (!prev) return prev;
      const items = prev.items.map((item) =>
        item.id === response.item.id ? response.item : item,
      );
      return recountToday({ ...prev, items, day_completed: response.day_completed });
    });
    void queryClient.invalidateQueries({ queryKey: traineeKeys.today() });
  };
}

export function useSaveTraineeCall() {
  const apply = useApplyItemResponse();
  return useMutation({
    mutationFn: ({ itemId, body }: { itemId: number; body: TraineeCallBody }) =>
      traineeApi.saveCall(itemId, body),
    onSuccess: apply,
  });
}

export function useSaveTraineeOutcome() {
  const apply = useApplyItemResponse();
  return useMutation({
    mutationFn: ({ itemId, body }: { itemId: number; body: TraineeOutcomeBody }) =>
      traineeApi.saveOutcome(itemId, body),
    onSuccess: apply,
  });
}

export function useTraineeOpenJobs(itemId: number | null) {
  return useQuery({
    queryKey: traineeKeys.openJobs(itemId ?? 0),
    queryFn: () => traineeApi.openJobs(itemId as number),
    enabled: itemId != null,
  });
}

export function useTraineeHandover() {
  return useMutation({
    mutationFn: ({ itemId, body }: { itemId: number; body: TraineeHandoverBody }) =>
      traineeApi.handover(itemId, body),
  });
}

export function useTraineeOverview() {
  return useQuery({
    queryKey: traineeKeys.overview(),
    queryFn: () => traineeApi.overview(),
  });
}

export function useUpdateTraineeProgram() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ userId, body }: { userId: number; body: TraineeProgramUpdate }) =>
      traineeApi.updateProgram(userId, body),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: traineeKeys.overview() }),
  });
}

export function useTraineeDecision() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ userId, body }: { userId: number; body: TraineeDecisionBody }) =>
      traineeApi.decision(userId, body),
    onSuccess: () => queryClient.invalidateQueries({ queryKey: traineeKeys.overview() }),
  });
}

export function useTraineeQualitySample(userId: number | null) {
  return useQuery({
    queryKey: traineeKeys.qualitySample(userId ?? 0),
    queryFn: () => traineeApi.qualitySample(userId as number),
    enabled: userId != null,
  });
}

export function useTraineeQualityVerdict(userId: number) {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({
      itemId,
      body,
    }: {
      itemId: number;
      body: { verdict: QualityVerdict; note: string };
    }) => traineeApi.qualityVerdict(itemId, body),
    onSuccess: (_data, { itemId, body }) => {
      queryClient.setQueryData<TraineeQualitySampleItem[]>(
        traineeKeys.qualitySample(userId),
        (prev) =>
          prev?.map((row) =>
            row.item_id === itemId ? { ...row, verdict: body.verdict, note: body.note } : row,
          ),
      );
      void queryClient.invalidateQueries({ queryKey: traineeKeys.overview() });
    },
  });
}

export function useTraineeRules() {
  return useQuery({ queryKey: traineeKeys.rules(), queryFn: () => traineeApi.rules() });
}

export function useSaveTraineeRules() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (body: TraineeRules) => traineeApi.saveRules(body),
    onSuccess: (saved) => {
      queryClient.setQueryData(traineeKeys.rules(), saved);
      void queryClient.invalidateQueries({ queryKey: ["trainee", "rules-preview"] });
      void queryClient.invalidateQueries({ queryKey: traineeKeys.overview() });
    },
  });
}

export function useTraineeRulesPreview(rules: TraineeRules | null) {
  return useQuery({
    queryKey: traineeKeys.rulesPreview(rules),
    queryFn: () => traineeApi.rulesPreview(rules),
    enabled: rules != null,
    placeholderData: (previous) => previous,
  });
}

File: backend/app/api/trainee.py
"""Praktykant — „Telefony na dziś” i panel Head of Recruitment (0374).

Trasy praktykanta stoją poza sekcjami (praktykant nie ma żadnej) na
``TraineeUser`` i widzą wyłącznie pozycje własnej listy; karta kandydata to
wąski odczyt z tej listy, nie profil. Trasy panelu — admin i Head of
Recruitment (``HeadOfRecruitmentPlus``). Kontrakt:
``docs/trainee-call-lists-contract.md``.
"""

from datetime import date
from typing import Literal, Optional

from fastapi import APIRouter, Depends, Query, Request
from pydantic import BaseModel, Field, field_validator
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import HeadOfRecruitmentPlus, TraineeUser
from app.api.section_access import PIPELINE_SECTION_DEPENDENCIES
from app.core.database import get_db
from app.core.scheduling import business_today
from app.services import trainee_call_list as lists
from app.services import trainee_program as program_service
from app.services import trainee_rules as rules_mod

router = APIRouter()


class MinRate(BaseModel):
    value: float = Field(gt=0, lt=1_000_000)
    unit: Literal["hour", "day", "month"] = "hour"


class CallFactsIn(BaseModel):
    b2b_willingness: Literal["b2b", "would_switch", "employment_only"]
    min_rate: Optional[MinRate] = None
    accepts_below_min_rate: Optional[bool] = None
    remote_modes: list[Literal["remote", "hybrid", "onsite"]] = Field(
        default_factory=list, max_length=3
    )
    max_onsite_days: Optional[int] = Field(default=None, ge=0, le=5)
    accepts_more_office_days: Optional[bool] = None
    office_cities: list[str] = Field(default_factory=list, max_length=10)
    work_time_preference: Optional[
        Literal["full_time_only", "also_part_time", "part_time_only"]
    ] = None
    availability: Optional[Literal["now", "within_1m", "within_3m", "later"]] = None
    open_to_offers: Optional[Literal["yes", "maybe", "no"]] = None
    wants: Optional[str] = Field(default=None, max_length=1000)

    @field_validator("office_cities")
    @classmethod
    def _cities(cls, value: list[str]) -> list[str]:
        out = []
        for city in value:
            clean = city.strip()[:80]
            if clean and clean not in out:
                out.append(clean)
        return out

    @field_validator("wants")
    @classmethod
    def _wants(cls, value: Optional[str]) -> Optional[str]:
        return value.strip() or None if value else None

    def to_facts(self) -> program_service.CallFacts:
        return program_service.CallFacts(
            b2b_willingness=self.b2b_willingness,
            min_rate_value=self.min_rate.value if self.min_rate else None,
            min_rate_unit=self.min_rate.unit if self.min_rate else "hour",
            accepts_below_min_rate=self.accepts_below_min_rate,
            remote_modes=tuple(dict.fromkeys(self.remote_modes)),
            max_onsite_days=self.max_onsite_days,
            accepts_more_office_days=self.accepts_more_office_days,
            office_cities=tuple(self.office_cities),
            work_time_preference=self.work_time_preference,
            availability=self.availability,
            open_to_offers=self.open_to_offers,
            wants=self.wants,
        )


class OutcomeIn(BaseModel):
    outcome: Literal["noanswer", "later", "wrong", "declined"]
    later_date: Optional[date] = None


class HandoverIn(BaseModel):
    job_id: int
    note: Optional[str] = Field(default=None, max_length=1000)


class ProgramIn(BaseModel):
    start_date: Optional[date] = None
    workdays: Optional[int] = Field(default=None, ge=1, le=250)
    daily_list_size: Optional[int] = Field(default=None, ge=1, le=300)


class DecisionIn(BaseModel):
    action: Literal["promote", "extend", "end"]
    role: Optional[Literal["sourcer", "recruiter"]] = None
    add_to_my_people: bool = True
    extend_days: int = Field(default=20, ge=1, le=120)


class QualityIn(BaseModel):
    verdict: Literal["ok", "issue"]
    note: Optional[str] = Field(default=None, max_length=1000)


# ── Praktykant ────────────────────────────────────────────────────────────


def _is_preview(request: Request) -> bool:
    """Admin w „podglądzie jako” — GET ma wtedy niczego nie zapisywać."""
    return getattr(request.state, "impersonator_id", None) is not None


@router.get("/today")
async def get_today(
    request: Request, user: TraineeUser, db: AsyncSession = Depends(get_db)
):
    return await program_service.today_view(db, user, read_only=_is_preview(request))


@router.post("/items/{item_id}/call")
async def save_call(
    item_id: int,
    body: CallFactsIn,
    user: TraineeUser,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.save_call(db, user, item_id, body.to_facts())


@router.post("/items/{item_id}/outcome")
async def save_outcome(
    item_id: int,
    body: OutcomeIn,
    user: TraineeUser,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.record_outcome(
        db, user, item_id, body.outcome, body.later_date
    )


@router.get("/items/{item_id}/open-jobs")
async def get_open_jobs(
    item_id: int, user: TraineeUser, db: AsyncSession = Depends(get_db)
):
    return await program_service.open_jobs(db, user, item_id)


@router.post("/items/{item_id}/handover")
async def post_handover(
    item_id: int,
    body: HandoverIn,
    user: TraineeUser,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.handover(db, user, item_id, body.job_id, body.note)


# ── Head of Recruitment i admin ───────────────────────────────────────────
# Sufit sekcji Pipeline (lustro „Moich ludzi” i propozycji) + rola.


@router.get("/overview", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def get_overview(
    _user: HeadOfRecruitmentPlus, db: AsyncSession = Depends(get_db)
):
    return await program_service.overview(db)


@router.put("/programs/{user_id}", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def put_program(
    user_id: int,
    body: ProgramIn,
    _user: HeadOfRecruitmentPlus,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.update_program(
        db,
        user_id,
        start_date=body.start_date,
        workdays=body.workdays,
        daily_list_size=body.daily_list_size,
    )


@router.post("/programs/{user_id}/decision", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def post_decision(
    user_id: int,
    body: DecisionIn,
    user: HeadOfRecruitmentPlus,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.decide(
        db,
        user,
        user_id,
        action=body.action,
        role=body.role,
        add_to_my_people=body.add_to_my_people,
        extend_days=body.extend_days,
    )


@router.get("/quality-sample", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def get_quality_sample(
    _user: HeadOfRecruitmentPlus,
    user_id: int = Query(...),
    db: AsyncSession = Depends(get_db),
):
    return await program_service.quality_sample(db, user_id)


@router.post("/quality-sample/{item_id}", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def post_quality(
    item_id: int,
    body: QualityIn,
    user: HeadOfRecruitmentPlus,
    db: AsyncSession = Depends(get_db),
):
    return await program_service.set_quality(db, user, item_id, body.verdict, body.note)


@router.get("/rules", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def get_rules(_user: HeadOfRecruitmentPlus, db: AsyncSession = Depends(get_db)):
    return await lists.load_rules(db)


@router.put("/rules", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def put_rules(
    body: dict,
    user: HeadOfRecruitmentPlus,
    db: AsyncSession = Depends(get_db),
):
    rules = await lists.save_rules(db, body, user_id=user.id)
    await db.commit()
    lists.reset_ranking_cache()
    return rules


@router.get("/rules/preview", dependencies=PIPELINE_SECTION_DEPENDENCIES)
async def get_rules_preview(
    request: Request,
    _user: HeadOfRecruitmentPlus,
    db: AsyncSession = Depends(get_db),
):
    """Wielkość puli przy regułach z formularza (niezapisane wartości w query
    nadpisują zapisane). Liczone świeżo — kilka sekund, w wątku, jedno
    liczenie na te same reguły i pamiętane chwilę (runda 6 audytu).
    Statystyki panelu zapisujemy tylko dla reguł zapisanych."""
    saved = await lists.load_rules(db)
    overrides = _query_rules(request.query_params)
    rules = rules_mod.normalize_rules({**saved, **overrides})
    stats = await lists.preview_pool_stats(db, rules, today=business_today())
    if rules == saved and not _is_preview(request):
        await lists.store_pool_stats(db, stats)
        await db.commit()
    return stats


def _query_rules(params) -> dict:
    out: dict = {}
    for key, raw in params.items():
        if key not in rules_mod.DEFAULT_RULES:
            continue
        if isinstance(rules_mod.DEFAULT_RULES[key], bool):
            if raw.lower() in ("true", "1"):
                out[key] = True
            elif raw.lower() in ("false", "0"):
                out[key] = False
        else:
            out[key] = raw
    return out
